### ISSUE TITLE
MCUSR is cleared in bootloader

### ISSUE THREAD

**Issue by GoogleCodeExporter** [Optiboot/optiboot#97]
```
I am running optiboot on an ATmega1284P and everything is working perfectly! 
But I would like to detect when a watchdog timer caused the reset, but due to 
this line: 
https://code.google.com/p/optiboot/source/browse/optiboot/bootloaders/optiboot/o
ptiboot.c#433 it is not possible at the moment, as the MCUSR register is set to 
0 just before the app starts.

What is the reason for doing this? Would it be possible for me just to clear 
everything but the WDRF register with no side effects?

Thanks a lot for all your work!
```

Original issue reported on code.google.com by `laus...@gmail.com` on 5 Mar 2014 at 9:44

**Comment by MarkG55** [comment]
I've modified the code to preserve most of the reset flags in MCUSR and created a new branch:
https://github.com/MarkG55/optiboot

**Comment by majekw** [comment]
I'm preparing fix for this. First part is ready for testing (probably doesn't work with virtual boot partition): https://github.com/Optiboot/optiboot/commit/14de60f0601a9b28f56e7f37ae0963db48fddb82 - comments welcome.

**Comment by WickedDevice** [comment]
Set the issue I posted that might rely on MCUSR being cleared on bootloader
exit?

**Comment by MarkG55** [comment]
Sorry - I'm not exactly clear on what you're asking?


On 30/05/2015 10:17 AM, WickedDevice wrote:
> Set the issue I posted that might rely on MCUSR being cleared on 
> bootloader
> exit?
> On May 29, 2015 7:07 PM, "majekw" <notifications@github.com> wrote:
>
> > I'm preparing fix for this. First part is ready for testing (probably
> > doesn't work with virtual boot partition): 14de60f
> > 
> <https://github.com/Optiboot/optiboot/commit/14de60f0601a9b28f56e7f37ae0963db48fddb82>
> > - comments welcome.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/Optiboot/optiboot/issues/97#issuecomment-106955667>.
> >
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/Optiboot/optiboot/issues/97#issuecomment-106966026>.
>


-- 
Regards,
Mark.

**Comment by vicatcu** [comment]
Sorry, I meant have a look at https://github.com/Optiboot/optiboot/issues/136

It might be a legitimate reason for clearing MCUSR from the bootloader.

**Comment by MarkG55** [comment]
I don't see that as being reason to clear MCUSR.  

With my changes to keep MCUSR unchanged, the bootloader will still only run when it's supposed to be run.  

I did also make another change to allow user code to jump to the bootloader, so it would be possible for an inadvertent jump to cause the bootloader to run, but that won't actually do any damage unless you're really unlucky and you're receiving data from the serial port which the bootloader interprets as commands to flash the user code area.  I personally don't think that this is a serious problem, but if you are concerned by it, you can remove/disable that part of my code.

The most likely scenario which could cause flash corruption due to an incorrect jump would be if the jump was to the middle of the boot loader.  Clearing MCUSR wouldn't help with this because for this invalid jump to be a problem, it must have bypassed the initial check of MCUSR anyway!  You could add in extra code to check MCUSR just before executing the actual flash operation, but that's only going to reduce the likelihood of the problem occurring without eliminating it completely.  Since the likelihood of this problem should already be very very small, I personally don't think it's worth worrying about.

**Comment by MarkG55** [comment]
majekw: If I'm reading it correctly, the code that you're linking to doesn't work as well as my changes - it looks like it still clears MCUSR unnecessarily under some circumstances...

**Comment by majekw** [comment]
MarkG55: my code clears MCUSR only when both EXTRF and WDRF are set. It occurs in one situation: watchdog triggered in bootloader code, so passing WDRF flag to application will mislead it, because it wasn't application WDT, and application shouldn't do anything like after WDT (for example skip some initializations). So in this case clearing at least WDRF is advisable.

Why I'm clearing more: if I clear only WDRF, on next watchdog reset in application EXTRF will be still set and bootloader could classify it as 'own' watchdog reset and don't pass WDRF back to application, so I need to clear both of them. Ok, I clear also rest of flags, but if they are still set, it's clear sign that application doesn't care about it :-)
And writing zero saves few bytes in code.

What are advantages of my code:
- it passes WDT flag to application when it was triggered by application
- stores original MCUSR in r2, so it's backward compatible with old Optiboot's way of passing MCUSR to application
- doesn't touch any registers except r1 and r2 and don't clobber memory or stack

About jumping to bootloader code - great idea (#52). My code do it also with cleared MCUSR:-) Only timeout isn't increasaed as in your code.

Did you checked difference in bootloader flash size before and after your modification?

**Comment by MarkG55** [comment]
Hi majekw,

If I'm reading your code correctly, if the user sketch doesn't clear MCUSR, your code won't run the bootloader after an external reset because the Power On Reset Flag will still be set.  I could see this being a serious problem for newbies writing their first sketch and then not being able to do any new firmware updates via the bootloader!

With my code, the way for the user code to distinguish a real watchdog reset is if MCUSR == 0, then a watchdog reset was the cause of the reset.  An external reset flag will be visible to the user code as normal with my code whereas it will be cleared with yours.  I prefer my method, but the end result is probably much the same either way as far as requiring changes to the end user code.

My code continues to store the original MCUSR value in r2 for backward compatibility, so there's no real difference there.

I haven't checked the difference in bootloader flash size - I don't think there was more than a few bytes difference with my changes though.  Even though you're using assembler, I think that your code would end up using more space than mine.  My code should only be doing a bitwise and operation and comparing it to a constant value before deciding on a single jump - yours has 3 conditional jumps.

A side effect of my code is that it's possible to flash the bootloader and then flash user code without needing to have the external reset line connected - after programming the bootloader with an In System Programmer, the boot loader runs continuously until either the board is powered off, or firmware is uploaded to the chip.

**Comment by majekw** [comment]
Hi MarkG55,

You clearly spotted a bug. You are perfectly right about PORF issue. Thanks! Fixed in https://github.com/Optiboot/optiboot/commit/0d1a8cf2805ea818ac131af6ec69ec31fb8b46a3
That's what code review is for :-)

If I read your code correctly, you pass to bootloader code only when WDRF=0 and EXTRF=1. For other combinations you always clear WDRF and jump to application.
I see one problem here: when application uses watchdog, it will never receive desired WDRF flag set.

About: MCUSR = ~(_BV(WDRF))
Is it safe to assume that bits in MCUSR could be only reset? I didn't checked it in real hardware, but I would be cautious about that. Most documentation uses MCUSR &= ~(_BV(WDRF)) construction.

**Comment by MarkG55** [comment]
You're correct that my code doesn't leave WDRF set, but if you read my comments above, the user code can still test for this - if MCUSR == 0, then the reset was due to the watchdog.  (I've actually got my application code checking all of the other flags first and then assuming a watchdog reset if it falls through - this way the code works the same even if there is no bootloader being used)

Because of the fact that Optiboot uses the watchdog, keeping the WDRF set isn't actually that useful and can actually cause problems - consider the following scenario with your code:

1.  User code running along normally.
2.  External reset occurs.
3.  Bootloader runs, times out with watchdog and controller resets.
4.  Bootloader runs user code without resetting WDRF.

Under the above circumstances, WDRF will be set even though it wasn't a true watchdog reset condition.  The user code could distinguish this by checking the status of EXTRF, but that's no better than having WDRF cleared anyway!

If the end user code doesn't use the watchdog and doesn't clear MCUSR, then the problem is even more serious - you can't disable the watchdog whilst WDRF is set (it's a safety feature) and the controller will just sit there resetting every 16 ms!

If I'm reading your latest code correctly, there is now also the problem that all other reset flags will be cleared every time.

Yes: you can only clear the flags in MCUSR, so there is no danger of my code setting any of the other flags - it should use less bytes of flash doing it my way.

**Comment by WestfW** [comment]
Wait...  too many cooks!
How does this compare to the last patch in the comments here https://github.com/Optiboot/optiboot/pull/125 which I think leaves R2 with an accurate reset cause, but still clears MCUSR.

**Comment by MarkG55** [comment]
I haven't tried running the code in that comment, but my expectation would be that with a true watchdog reset, the reset cause wouldn't be placed into r2 - r2 would would then keep whatever value was in it when the bootloader started.  Assuming that the user code modifies r2, it would then contain effectively random data.

Another thing that I'm not sure about is whether the value in r2 would actually survive a watchdog reset?  If it doesn't, then that code wouldn't work at all...

**Comment by WestfW** [comment]
The reason to clear MCUSR is that its bits are sticky across resets, so you wind up with multiple bits being set, and don't know which one is relevant.  From tests, it looks like there are normal cases where multiple bits get set as well.  (There's a simple test script here, now: https://github.com/Optiboot/optiboot/blob/master/optiboot/examples/test_reset/test_reset.ino )  normal C startup code, and Arduino startup code, does NOT reset MCUSR, so if Optiboot doesn't clear it you'll end up in situation where the bootload doesn't run any more, because it's doing 'fast start' for all non-RSTRF startup reasons (Hmm.  That may be a bug :-( )

Maybe a step backward?  What are the "interesting cases" to a user application?
1) Power on.  Typically BORF+PORF will be set, and the sketch is run "instantly."  Apparently, sometimes RSTRF is set as well (depending on hardware?)(which is a problem.)
2) Reset signal.  The tough case - optiboot runs and eventually starts the application using WDT.
3) WDT in application - it would be nice to know if this happens.  The streef pull request (and my modification of this) does this via r2.
4) actual brownout.

**Comment by majekw** [comment]
WestfW:
For me, perfect bootloader should be transparent for application.
Sadly, Optiboot isn't as it destroys MCUSR. It doesn't matter if we preserve status in r2 or somewhere else - application expects some flags in MCUSR. So, in my opinion we should preserve contents of MCUSR as much as possible. And as you see from discussion here, it's mostly possible. That's why I think that #125 doesn't try to resolve real problem - it just tries to improve workaround (in some weird way).

**Comment by majekw** [comment]
This is my code translated to C, so you could get what I was thinking aboout:
    c = MCUSR
    if (c & _BV(WDRF)) {
        if (c & _BV(EXTRF) MCUSR=0;
        appStart();
    }
    if (c & _BV(EXTRF) appStart();
    ... procedd to bootloader

**Comment by majekw** [comment]
So, to summmarize, assuming for a while that bootloader don't do any harm to MCUSR:
1. Bootloader should be entered when EXTRST is set - we can't change it
2. Bootlader shouldn't be entered where both EXTRST and WDTF are set, as it's a situation where right after bootloader timeout or first 'start' to application which occurs using watchdog.
3. Bootloader in case 2. shouldn't pass WDTF flag to application as it wasn't application watchdog event!

**Comment by majekw** [comment]
So, my code check first for watchdog. If watchdog flag is set, we have 2 options:
1. Application watchdog triggered, so we should pass it as is
2. Bootloader watchdog triggered, so we should clear WDTF to not confuse application.
Difference between these 2 situations is EXTRF. So, why I clear more?
If in second scenario I whould leave EXTRF and applications doesn't clear all flags, but only WDTF (as documentation suggest), next reset trigger (for example from BOD) could be get by bootloader as request to step in, which is not true.
Ok, I clear also BODF and POWF, but I think that in that specific scenario (WDT reset from bootloader) we could sacrifice some information :-) I can change it to something like MCUSR &= ~(_BV(WDRF)|_BV(EXTRF)) to be almost perfect, but at cost of another 4B and one additional register destroyed.

After dealing with watchdog exception I just checks for EXTRF and jump into bootloader if it exist.
Regarding flags: there shouldn't be false false positives from ored flags, because code clears EXTRF (with oher flags) on first timeout/jump to application.

**Comment by majekw** [comment]
WestfW: you are right about 'think first, code later' :-)
I think we could discuss desired flow in new issue #140, and then get back here to code it.

**Comment by MarkG55** [comment]
If you want to keep WDRF set when it's caused by an application reset, you could modify my code as follows:

```
  ch = MCUSR;

  if (ch == 0) {
      WaitTime = WATCHDOG_8S;   // If no Reset Flags are set, it must mean that the user code jumped to the boot loader!  Wait 8 seconds for the firmware to be updated!
  } else {
      if ((ch & (_BV(WDRF) | _BV(EXTRF))) != _BV(EXTRF)) {    // To run the boot loader, External Reset Flag must be set and the Watchdog Flag MUST be cleared!  Otherwise jump straight to user code.
        if (ch & _BV(EXTRF)) 
          MCUSR = ~(_BV(WDRF));  // Clear WDRF because it was actually caused by bootloader
        appStart(ch);
      }
  }
```

Of course, this does add a few bytes to the size of the bootloader, but some people may think that it's worthwhile...  Note that if the user code doesn't clear EXTRF but does use the watchdog, it won't see WDRF set after an external reset has occurred - will be a more serious problem on hardware that causes EXTRF to be set during power on.

Please note that unlike my original code, I haven't actually tested this change.

I'll include my code which translates the flags in MCUSR to architecture agnostic values that I've defined elsewhere.  This code works the same with no bootloader, the original Arduino bootloader, my modified optiboot code and should also still work with the above modification.  It relies on the fact that I clear MCUSR in my user code after saving it to a temporary variable.

```
#ifdef __AVR__
byte GetResetCode(byte SavedMCUSR)
{
  if (SavedMCUSR & _BV(PORF))
    return Startup_PowerOn;

  if (SavedMCUSR & _BV(EXTRF))
    return Startup_ExternalReset;

  if (SavedMCUSR & _BV(BORF))
    return Startup_BrownOutReset;

  return Startup_WatchDogReset;
}

#endif
```

**Comment by MarkG55** [comment]
One possible problem with keeping WDRF set is that if the end user code enables the watchdog, but doesn't clear the flag after a reset, an external reset won't run the bootloader.  Resetting it for a second time will however cause the bootloader to run correctly - as long as the board doesn't end up in a constant watchdog reset loop.

If a newbie enables the watchdog and doesn't clear WDRF at the start of their code and they don't reprogram the watchdog timer soon enough, the controller will just sit there resetting itself constantly once a watchdog event has occurred and it won't be possible to reflash new code without power cycling the board.

Even worse, if we keep WDRF set, a newbie could soft brick their device by enabling the watchdog straight away and then not reset the watchdog soon enough - would require the "hold reset while powering on" trick as mentioned here:
https://github.com/Optiboot/optiboot/issues/140

With the sample code here:
http://forum.arduino.cc/index.php?topic=150369.msg1129533#msg1129533

My original code will always start the bootloader with an external reset, with the change to keep WDRF, it will be harder to recover from - requiring either 2 resets within about 1 second, or the "hold reset while powering on" trick.

**Comment by MarkG55** [comment]
I was going to suggest another alteration to my changes, but thinking about it a bit more, I don't think it's actually necessary - with my latest version, the chances of soft bricking the device because the user enables the watchdog but doesn't clear WDRF isn't as great as I previously thought.

As long as the user code either clears MCUSR completely, or doesn't clear it at all, then this code should be fine even if the user tries to soft brick the device with the watchdog timer.

If the user code clears EXTRF, but doesn't clear WDRF, then it will still be possible to soft brick the device - requiring the "hold reset during power on" trick to recover.  The chances of a user doing this though are pretty slim, so I don't think it's a big problem.

If the user code uses the watchdog but doesn't clear EXTRF, then it won't see WDRF come through after an external reset.  Again, I don't think that this is a serious problem since the user is unlikely to do this and the consequences are relatively minor anyway.

**Comment by WestfW** [comment]
Fixed in V7.0
